File: tests/cpp/host_overlay_clip_test.cpp
#include "render/host_overlay/clip.h"
#include "render/host_overlay/software/renderer.h"
#include "support/host_overlay_fixture.h"
#include "spec/bmsx/model.h"
#include <algorithm>
#include <iostream>
#include <stdexcept>
#include <vector>

using namespace bmsx;

static void require(bool condition, const char* message) {
	if (!condition) throw std::runtime_error(message);
}

int main() {
	HostOverlayClipState scissor;
	const HostOverlayClipRect clip{11, 9, 43, 34};
	scissor.reset(64, 48, 160, 120);
	scissor.set(clip);
	require(scissor.left == 27 && scissor.top == 22 && scissor.right == 107 && scissor.bottom == 85, "logical scissor mapping");
	scissor.set({-30, -20, 100, 100});
	require(scissor.left == 0 && scissor.top == 0 && scissor.right == 160 && scissor.bottom == 120, "scissor target intersection");
	scissor.set({80, 30, 90, 20});
	require(scissor.left == 160 && scissor.top == 75 && scissor.right == 160 && scissor.bottom == 75, "empty scissor intersection");

	SoftwareBackend backend(64, 48, PSX_MACHINE_SPEC.gxGpuVramBytes);
	backend.resizePresentationTarget(64, 48);
	Host2DPipelineState state;
	test::HostOverlayFixture fixture;
	std::vector<u32> reference(64 * 48);
	for (const HostOverlayTransform transform : {IDENTITY_HOST_OVERLAY_TRANSFORM, HostOverlayTransform{0.5F, 3, 2}, HostOverlayTransform{2, 3, 2}}) {
		for (size_t index = 0; index < fixture.kinds.size(); ++index) {
			std::fill_n(backend.framebuffer(), reference.size(), 0);
			beginHostOverlaySoftware(backend, state);
			renderHost2DEntrySoftware(backend, Host2DKind::Transform, {.transform = &transform});
			renderHost2DEntrySoftware(backend, fixture.kinds[index], fixture.refs[index]);
			std::copy_n(backend.framebuffer(), reference.size(), reference.begin());
			std::fill_n(backend.framebuffer(), reference.size(), 0);
			renderHost2DEntrySoftware(backend, Host2DKind::Clip, {.clip = &clip});
			renderHost2DEntrySoftware(backend, fixture.kinds[index], fixture.refs[index]);
			int lit = 0;
			for (int y = 0; y < 48; ++y) {
				for (int x = 0; x < 64; ++x) {
					const bool inside = x >= clip.left && x < clip.right && y >= clip.top && y < clip.bottom;
					const size_t offset = static_cast<size_t>(y * 64 + x);
					require(backend.framebuffer()[offset] == (inside ? reference[offset] : 0), "clipping must crop, not reposition or resample");
					lit += backend.framebuffer()[offset] != 0;
				}
			}
			require(lit > 0, "fixture must have partially visible pixels");
			endHostOverlaySoftware(backend);
			require(backend.hostOverlayClip.left == 0 && backend.hostOverlayClip.right == 64, "pass end releases the scissor");
		}
	}
	fixture.glyphs.x = 2;
	fixture.glyphs.y = 3;
	fixture.glyphs.items = {"A\tB\nC"};
	fixture.glyphs.item_end = 5;
	const HostOverlayTransform doubled{2, 0, 0};
	std::fill_n(backend.framebuffer(), reference.size(), 0);
	beginHostOverlaySoftware(backend, state);
	require(backend.hostOverlayTransform.scale == 1 && backend.hostOverlayTransform.offsetX == 0, "new pass resets transform");
	renderHost2DEntrySoftware(backend, Host2DKind::Glyphs, {.glyphs = &fixture.glyphs});
	std::copy_n(backend.framebuffer(), reference.size(), reference.begin());
	std::fill_n(backend.framebuffer(), reference.size(), 0);
	renderHost2DEntrySoftware(backend, Host2DKind::Transform, {.transform = &doubled});
	renderHost2DEntrySoftware(backend, Host2DKind::Glyphs, {.glyphs = &fixture.glyphs});
	for (int y = 0; y < 48; ++y) {
		for (int x = 0; x < 64; ++x) {
			require(backend.framebuffer()[y * 64 + x] == reference[(y / 2) * 64 + x / 2], "2x glyph geometry must reproduce each texel including tabs, newlines and backgrounds");
		}
	}
	endHostOverlaySoftware(backend);
	std::cout << "HOST-OVERLAY-CLIP:PASS\n";
}

File: tests/lua/host_overlay_clip.test.ts
import assert from 'node:assert/strict';
import test from 'node:test';
import { HostOverlayClipState, type HostOverlayClipRect } from '../../machine/ts/render/host_overlay/clip';
import { Host2DKind } from '../../machine/ts/render/host_overlay/commands';
import { HostOverlayQuadStream } from '../../machine/ts/render/host_overlay/quad_stream';
import { beginHeadlessHost2D, renderHeadlessHost2DEntry } from '../../machine/ts/render/headless/host_2d';
import { LAYER_2D_IDE } from '../../machine/ts/render/shared/layers';
import { RectRenderKind } from '../../machine/ts/render/shared/submissions';
import { createHostOverlayFixture } from '../helpers/host_overlay';
import { hostOverlayPrimitives } from '../helpers/host_overlay_primitives';

const full = { left: 0, top: 0, right: 64, bottom: 48 };
const clip = { left: 11, top: 9, right: 43, bottom: 34 };

test('host scissor maps logical coordinates once, bounds the target and preserves empty intersections', () => {
	const state = new HostOverlayClipState();
	state.reset(64, 48, 160, 120);
	state.set(clip);
	assert.deepEqual([state.left, state.top, state.right, state.bottom], [27, 22, 107, 85]);
	state.set({ left: -30, top: -20, right: 100, bottom: 100 });
	assert.deepEqual([state.left, state.top, state.right, state.bottom], [0, 0, 160, 120]);
	state.set({ left: 80, top: 30, right: 90, bottom: 20 });
	assert.deepEqual([state.left, state.top, state.right, state.bottom], [160, 75, 160, 75]);
	state.reset(64, 48, 64, 48);
	assert.deepEqual([state.left, state.top, state.right, state.bottom], [0, 0, 64, 48]);
});

test('nested clip publication does not alias a reused stack depth or the next frame', () => {
	const { presenter, queue, renderer } = createHostOverlayFixture(64, 48);
	const draw = () => {
		renderer.beginFrame(presenter);
		renderer.pushClipRect(-2, 5, 50, 40);
		renderer.pushClipRect(10, 0, 60, 30);
		renderer.popClipRect();
		renderer.pushClipRect(20, 8, 30, 12);
		renderer.popClipRect();
		renderer.popClipRect();
		renderer.endFrame();
		return queue.consumeOverlayFrame();
	};
	const frame = draw();
	assert.deepEqual(frame.commandRefs.slice(0, frame.commandCount), [
		{ left: 0, top: 5, right: 50, bottom: 40 },
		{ left: 10, top: 5, right: 50, bottom: 30 },
		{ left: 0, top: 5, right: 50, bottom: 40 },
		{ left: 20, top: 8, right: 30, bottom: 12 },
		{ left: 0, top: 5, right: 50, bottom: 40 }, full,
	]);
	const refs = frame.commandRefs.slice(0, frame.commandCount);
	const next = draw();
	assert.notEqual(frame.commandRefs, next.commandRefs);
	assert.deepEqual(frame.commandRefs.slice(0, frame.commandCount), refs);
	const reused = draw();
	for (let index = 0; index < refs.length; index += 1) assert.equal(reused.commandRefs[index], refs[index]);
});

test('quad batches split only at a changed clip with preceding geometry and retain their storage', () => {
	const stream = new HostOverlayQuadStream();
	const rect = { kind: RectRenderKind.Fill, area: { ...full, z: 0 }, color: 0xffffffff, layer: LAYER_2D_IDE };
	const draw = () => {
		stream.reset(64, 48);
		stream.appendEntry(Host2DKind.Rect, rect);
		stream.appendEntry(Host2DKind.Clip, clip);
		stream.appendEntry(Host2DKind.Clip, { ...clip });
		stream.appendEntry(Host2DKind.Rect, rect);
		stream.appendEntry(Host2DKind.Clip, full);
		stream.appendEntry(Host2DKind.Rect, rect);
	};
	draw();
	assert.equal(stream.batchCount, 3);
	assert.deepEqual(stream.batches.map(batch => batch.start), [0, 1, 2]);
	const batches = stream.batches.slice();
	const floats = stream.floatData;
	for (let frame = 0; frame < 100; frame += 1) draw();
	assert.equal(stream.floatData, floats);
	for (let index = 0; index < batches.length; index += 1) assert.equal(stream.batches[index], batches[index]);
});


for (const [name, kind, command] of hostOverlayPrimitives) {
	test(`software ${name}: clipping crops the original raster and never repositions geometry or UVs`, () => {
		const { backend } = createHostOverlayFixture(64, 48);
		const context = backend.hostOverlayContext;
		const target = backend.framebufferPixels;
		const draw = (bounds: HostOverlayClipRect) => {
			target.fill(0);
			beginHeadlessHost2D(context, target, 64, 48);
			renderHeadlessHost2DEntry(context, Host2DKind.Clip, bounds);
			renderHeadlessHost2DEntry(context, kind, command);
		};
		draw(full);
		const reference = target.slice();
		draw(clip);
		let lit = 0;
		for (let y = 0; y < 48; y += 1) {
			for (let x = 0; x < 64; x += 1) {
				const inside = x >= clip.left && x < clip.right && y >= clip.top && y < clip.bottom;
				for (let channel = 0; channel < 4; channel += 1) {
					const offset = (y * 64 + x) * 4 + channel;
					assert.equal(target[offset], inside ? reference[offset] : 0, `${name}: ${x},${y}:${channel}`);
					lit += target[offset] !== 0 ? 1 : 0;
				}
			}
		}
		assert.ok(lit > 0, 'fixture must contain partially visible pixels');
		beginHeadlessHost2D(context, target, 64, 48);
		assert.equal(context.clip.right, 64, 'a new lane resets its scissor');
	});
}
